# HG changeset patch
# User Betterbird <betterbird@betterbird.eu>
# Date 1744491780 -7200
# Parent  3fd46290910c6c6ee73aa404388b6c06647494e4
Feature: Account colours, part 1: Account Manager and folder tree.

See https://bugzilla.mozilla.org/show_bug.cgi?id=1482172 and
https://addons.thunderbird.net/en-US/thunderbird/addon/account-colors/
* * *
Feature: Account colours, part 1: Adjustments

diff --git a/mail/base/content/about3Pane.js b/mail/base/content/about3Pane.js
--- a/mail/base/content/about3Pane.js
+++ b/mail/base/content/about3Pane.js
@@ -66,8 +66,11 @@ const { ThreadPaneColumns } = ChromeUtil
 
 // As defined in nsMsgDBView.h.
 const MSG_VIEW_FLAG_DUMMY = 0x20000000;
 
+const SPECIALUSE_NOT_OUTBOX =
+  Ci.nsMsgFolderFlags.SpecialUse & ~Ci.nsMsgFolderFlags.Queue;
+
 /**
  * The TreeListbox widget that displays folders.
  */
 var folderTree;
@@ -1280,8 +1283,10 @@ var folderPane = {
             const folderRow = folderPane._createFolderRow(
               this.name,
               childFolder
             );
+            // This is already colored by its server.
+            folderRow.style.backgroundColor = "";
             serverRow.insertChildInOrder(folderRow);
             folderPane._addSubFolders(childFolder, folderRow, this.name);
             return;
           }
@@ -1292,8 +1297,10 @@ var folderPane = {
           this.containerList,
           childFolder,
           this.name
         );
+        // This is already colored by its server.
+        folderRow.style.backgroundColor = "";
         folderPane._addSubFolders(childFolder, folderRow, this.name);
       },
 
       removeFolder(parentFolder, childFolder) {
@@ -2720,8 +2727,33 @@ var folderPane = {
       row.setServerName(name);
     }
   },
 
+  changeServerColor(folder) {
+    const color = folder.server.RGBAcolor;
+    for (const row of folderTree.querySelectorAll(
+      `li[data-server-key="${folder.server.key}"]`
+    )) {
+      const f = MailServices.folderLookup.getFolderForURL(row.uri);
+      // Put color only on folders in smart mode.
+      if (row.id.startsWith("smart-")) {
+        // Only put color on top Inbox, etc., not sub-folders.
+        if (f.flags & SPECIALUSE_NOT_OUTBOX || f.isServer) {
+          // If the parent already has the style, clear it.
+          if (row.parentElement?.closest("li")?.style.backgroundColor) {
+            row.style.backgroundColor = "";
+          } else {
+            row.style.backgroundColor = color;
+          }
+        }
+      } else if (row.id.startsWith("recent-")) {
+        row.style.backgroundColor = color;
+      } else if (f.isServer) {
+        row.style.backgroundColor = color;
+      }
+    }
+  },
+
   /**
    * Update the UI widget to reflect the real folder size when the "FolderSize"
    * property changes.
    *
@@ -7111,8 +7143,13 @@ var folderListener = {
         if (folder.isServer) {
           folderPane.changeServerName(folder, newValue);
         }
         break;
+      case "Color":
+        if (folder.isServer) {
+          folderPane.changeServerColor(folder);
+        }
+        break;
     }
   },
   onFolderIntPropertyChanged(folder, property, oldValue, newValue) {
     switch (property) {
diff --git a/mail/base/content/widgets/folder-tree-row.mjs b/mail/base/content/widgets/folder-tree-row.mjs
--- a/mail/base/content/widgets/folder-tree-row.mjs
+++ b/mail/base/content/widgets/folder-tree-row.mjs
@@ -11,8 +11,11 @@ ChromeUtils.defineESModuleGetters(lazy, 
   MailServices: "resource:///modules/MailServices.sys.mjs",
   XULStoreUtils: "resource:///modules/XULStoreUtils.sys.mjs",
 });
 
+const SPECIALUSE_NOT_OUTBOX =
+  Ci.nsMsgFolderFlags.SpecialUse & ~Ci.nsMsgFolderFlags.Queue;
+
 /**
  * Represents a single row in the folder tree. The row can be for a server or
  * a folder. Use `folderPane._createServerRow` or `folderPane._createFolderRow`
  * to create rows.
@@ -271,8 +274,12 @@ class FolderTreeRow extends HTMLLIElemen
       this.unreadCount = server.rootFolder.getNumUnread(isCollapsed);
       this.totalCount = server.rootFolder.getTotalMessages(isCollapsed);
     }
     this.setFolderPropertiesFromFolder(server.rootFolder);
+    const color = server.RGBAcolor;
+    if (color) {
+      this.style.backgroundColor = color;
+    }
   }
 
   /**
    * Set some properties based on the folder for this row.
@@ -301,8 +308,30 @@ class FolderTreeRow extends HTMLLIElemen
       this.classList.add("noselect-folder");
     } else {
       this.setAttribute("draggable", "true");
     }
+    const color = folder.server.RGBAcolor;
+    if (!color) {
+      return;
+    }
+    if (
+      this.id.startsWith("smart-") &&
+      // Only put color on top Inbox, etc., not sub-folders.
+      // There is a problem here were special folders which are not part
+      // of a unified folder get colored twice, once as special folder,
+      // then as part of a server.
+      // The style is removed later, see comment:
+      // "This is already colored by its server".
+      folder.flags & SPECIALUSE_NOT_OUTBOX
+    ) {
+      this.style.backgroundColor = color;
+    } else if (this.id.startsWith("recent-")) {
+      this.style.backgroundColor = color;
+    } else if (folderPane._isCompact && this.id.startsWith("unread-")) {
+      this.style.backgroundColor = color;
+    } else if (folderPane._isCompact && this.id.startsWith("favorite-")) {
+      this.style.backgroundColor = color;
+    }
   }
 
   /**
    * Update new message state of the row.
diff --git a/mail/modules/AccountManagerUtils.sys.mjs b/mail/modules/AccountManagerUtils.sys.mjs
--- a/mail/modules/AccountManagerUtils.sys.mjs
+++ b/mail/modules/AccountManagerUtils.sys.mjs
@@ -8,9 +8,9 @@
 
 import { FolderTreeProperties } from "resource:///modules/FolderTreeProperties.sys.mjs";
 
 export class AccountManagerUtils {
-  defaultServerColor = "#2493ef";
+  defaultServerColor = "#000000";
 
   /**
    * The account currently being edited.
    *
@@ -23,13 +23,9 @@ export class AccountManagerUtils {
    *
    * @type {string}
    */
   get serverColor() {
-    return (
-      FolderTreeProperties.getColor(
-        this.#account.incomingServer.rootFolder.URI
-      ) || this.defaultServerColor
-    );
+    return this.#account.incomingServer.color || this.defaultServerColor;
   }
 
   /**
    * @param {nsIMsgAccount} account
diff --git a/mailnews/base/prefs/content/AccountManager.js b/mailnews/base/prefs/content/AccountManager.js
--- a/mailnews/base/prefs/content/AccountManager.js
+++ b/mailnews/base/prefs/content/AccountManager.js
@@ -1884,8 +1884,12 @@ var gAccountTree = {
       // Add icons based on account type.
       if (server) {
         treeitem.dataset.serverKey = server.key;
         treeitem.classList.add("serverType-" + server.type);
+        const color = server.RGBAcolor;
+        if (color) {
+          treeitem.style.backgroundColor = color;
+        }
         if (server.isSecure) {
           treeitem.classList.add("isSecure");
         }
         // For IM accounts, we can try to fetch a protocol specific icon.
diff --git a/mailnews/base/prefs/content/am-main.js b/mailnews/base/prefs/content/am-main.js
--- a/mailnews/base/prefs/content/am-main.js
+++ b/mailnews/base/prefs/content/am-main.js
@@ -78,8 +78,22 @@ function serverPrettyNameOnBlur(event) {
   parent.setAccountLabel(gAccount.key, event.target.value);
   setAccountTitle();
 }
 
+function serverColorOnChange(event, reset = false) {
+  const row = parent.document.getElementById(gAccount.key);
+  if (row) {
+    if (reset) {
+      gAccount.incomingServer.color = "";
+      row.style.backgroundColor = "";
+      return;
+    }
+    const color = document.getElementById("serverColor").value;
+    gAccount.incomingServer.color = color;
+    row.style.backgroundColor = gAccount.incomingServer.RGBAcolor;
+  }
+}
+
 /**
  * Update an account's main settings title with the account name if applicable.
  */
 function setAccountTitle() {
@@ -97,17 +111,20 @@ function setAccountTitle() {
 function setServerColor() {
   const colorInput = document.getElementById("serverColor");
   colorInput.value = AMUtils.serverColor;
 
-  colorInput.addEventListener("input", event =>
-    AMUtils.previewServerColor(event.target.value)
-  );
-  colorInput.addEventListener("change", event =>
-    AMUtils.updateServerColor(event.target.value)
-  );
-  document
-    .getElementById("resetColor")
-    .addEventListener("click", () => resetServerColor());
+  colorInput.addEventListener("input", event => {
+    AMUtils.previewServerColor(event.target.value);
+    serverColorOnChange(event);
+  });
+  colorInput.addEventListener("change", event => {
+    AMUtils.updateServerColor(event.target.value);
+    serverColorOnChange(event);
+  });
+  document.getElementById("resetColor").addEventListener("click", event => {
+    resetServerColor();
+    serverColorOnChange(event, true);
+  });
 }
 
 function resetServerColor() {
   document.getElementById("serverColor").value = AMUtils.defaultServerColor;
diff --git a/mailnews/base/prefs/content/am-main.xhtml b/mailnews/base/prefs/content/am-main.xhtml
--- a/mailnews/base/prefs/content/am-main.xhtml
+++ b/mailnews/base/prefs/content/am-main.xhtml
@@ -87,9 +87,15 @@
             for="serverColor"
             data-l10n-id="account-settings-color-label"
           />
           <div class="input-container">
-            <input id="serverColor" type="color" class="input-inline" />
+            <input
+              id="serverColor"
+              type="color"
+              wsm_persist="true"
+              class="input-inline"
+              prefstring="mail.server.%serverkey%.color"
+            />
             <button
               id="resetColor"
               type="button"
               data-l10n-id="account-settings-reset-color-button"
diff --git a/mailnews/base/prefs/content/am-serverwithnoidentities.js b/mailnews/base/prefs/content/am-serverwithnoidentities.js
--- a/mailnews/base/prefs/content/am-serverwithnoidentities.js
+++ b/mailnews/base/prefs/content/am-serverwithnoidentities.js
@@ -100,21 +100,38 @@ function onSave() {
     .getElementById("server.storeContractID")
     .setAttribute("value", storeContractID);
 }
 
+function serverColorOnChange(event, reset = false) {
+  const row = parent.document.getElementById(gAccount.key);
+  if (row) {
+    if (reset) {
+      gAccount.incomingServer.color = "";
+      row.style.backgroundColor = "";
+      return;
+    }
+    const color = document.getElementById("serverColor").value;
+    gAccount.incomingServer.color = color;
+    row.style.backgroundColor = gAccount.incomingServer.RGBAcolor;
+  }
+}
+
 function setServerColor() {
   const colorInput = document.getElementById("serverColor");
   colorInput.value = AMUtils.serverColor;
 
-  colorInput.addEventListener("input", event =>
-    AMUtils.previewServerColor(event.target.value)
-  );
-  colorInput.addEventListener("change", event =>
-    AMUtils.updateServerColor(event.target.value)
-  );
-  document
-    .getElementById("resetColor")
-    .addEventListener("click", () => resetServerColor());
+  colorInput.addEventListener("input", event => {
+    AMUtils.previewServerColor(event.target.value);
+    serverColorOnChange(event);
+  });
+  colorInput.addEventListener("change", event => {
+    AMUtils.updateServerColor(event.target.value);
+    serverColorOnChange(event);
+  });
+  document.getElementById("resetColor").addEventListener("click", event => {
+    resetServerColor();
+    serverColorOnChange(event, true);
+  });
 }
 
 function resetServerColor() {
   document.getElementById("serverColor").value = AMUtils.defaultServerColor;
diff --git a/mailnews/base/prefs/content/am-serverwithnoidentities.xhtml b/mailnews/base/prefs/content/am-serverwithnoidentities.xhtml
--- a/mailnews/base/prefs/content/am-serverwithnoidentities.xhtml
+++ b/mailnews/base/prefs/content/am-serverwithnoidentities.xhtml
@@ -85,9 +85,15 @@
             for="serverColor"
             data-l10n-id="account-settings-color-label"
           />
           <div class="input-container">
-            <input id="serverColor" type="color" class="input-inline" />
+            <input
+              id="serverColor"
+              type="color"
+              wsm_persist="true"
+              class="input-inline"
+              prefstring="mail.server.%serverkey%.color"
+            />
             <button
               id="resetColor"
               type="button"
               data-l10n-id="account-settings-reset-color-button"
diff --git a/mailnews/base/public/nsIMsgIncomingServer.idl b/mailnews/base/public/nsIMsgIncomingServer.idl
--- a/mailnews/base/public/nsIMsgIncomingServer.idl
+++ b/mailnews/base/public/nsIMsgIncomingServer.idl
@@ -475,8 +475,16 @@ interface nsIMsgIncomingServer : nsISupp
   attribute long incomingDuplicateAction;
 
   // check if new hdr is a duplicate of a recently arrived header
   boolean isNewHdrDuplicate(in nsIMsgDBHdr aNewHdr);
+
+  /**
+   * Color for the server in the folder pane.
+   * color format: CSS format: #FF0000.
+   * RGBAcolor format: CSS format: rgba(r, g, b, t).
+   */
+  attribute ACString color;
+  readonly attribute ACString RGBAcolor;
 };
 
 %{C++
 /*
diff --git a/mailnews/base/src/MsgIncomingServer.sys.mjs b/mailnews/base/src/MsgIncomingServer.sys.mjs
--- a/mailnews/base/src/MsgIncomingServer.sys.mjs
+++ b/mailnews/base/src/MsgIncomingServer.sys.mjs
@@ -720,8 +720,40 @@ export class MsgIncomingServer {
     }
     return this.authMethod != Ci.nsMsgAuthMethod.OAuth2;
   }
 
+  get color() {
+    return this.getStringValue("color");
+  }
+
+  set color(value) {
+    this.setStringValue("color", value);
+    this.rootFolder.NotifyPropertyChanged(
+      "Color",
+      "", // Old value doesn't matter.
+      value
+    );
+  }
+
+  get RGBAcolor() {
+    const color = this.getStringValue("color");
+    // Consider #000000 as not set. Reject bogus values.
+    if (
+      !color ||
+      color == "#000000" ||
+      !color.startsWith("#") ||
+      color.length != 7
+    ) {
+      return "";
+    }
+
+    // Turn #RRGGBB into rgba(rr, gg, bb, tt)
+    const r = parseInt(color.substring(1, 3), 16);
+    const g = parseInt(color.substring(3, 5), 16);
+    const b = parseInt(color.substring(5, 7), 16);
+    return `rgba(${r}, ${g}, ${b}, 0.2)`;
+  }
+
   getStringValue(prefName) {
     try {
       return this._prefs.getStringPref(prefName);
     } catch (e) {
diff --git a/mailnews/base/src/nsMsgIncomingServer.cpp b/mailnews/base/src/nsMsgIncomingServer.cpp
--- a/mailnews/base/src/nsMsgIncomingServer.cpp
+++ b/mailnews/base/src/nsMsgIncomingServer.cpp
@@ -41,8 +41,9 @@
 #include "nsIObserverService.h"
 #include "mozilla/Unused.h"
 #include "nsIUUIDGenerator.h"
 #include "nsIArray.h"
+#include "nsPrintfCString.h"
 #ifdef MOZ_PANORAMA
 #  include "nsIDatabaseCore.h"
 #  include "nsIFolder.h"
 #  include "nsIFolderDatabase.h"
@@ -2025,4 +2026,53 @@ NS_IMETHODIMP nsMsgIncomingServer::IsNew
     }
   }
   return NS_OK;
 }
+
+NS_IMETHODIMP
+nsMsgIncomingServer::GetColor(nsACString& color) {
+  return GetStringValue("color", color);
+}
+
+NS_IMETHODIMP
+nsMsgIncomingServer::SetColor(const nsACString& color) {
+  nsresult rv = SetStringValue("color", color);
+  NS_ENSURE_SUCCESS(rv, rv);
+  nsCOMPtr<nsIMsgFolder> rootFolder;
+  GetRootFolder(getter_AddRefs(rootFolder));
+  if (rootFolder) {
+    // Old value doesn't matter.
+    rootFolder->NotifyPropertyChanged("Color"_ns, EmptyCString(), color);
+  }
+  return NS_OK;
+}
+
+NS_IMETHODIMP
+nsMsgIncomingServer::GetRGBAcolor(nsACString& aColor) {
+  aColor.Truncate();
+  nsCString color;
+  nsresult rv = GetStringValue("color", color);
+  if (NS_FAILED(rv)) return NS_OK;
+
+  // Consider #000000 as not set. Reject bogus values.
+  if (color.IsEmpty() || color.EqualsLiteral("#000000") || color[0] != '#' ||
+      color.Length() != 7) {
+    return NS_OK;
+  }
+
+  // Turn #RRGGBB into rgba(rr, gg, bb, tt)
+  char colorBytes[3];
+  colorBytes[2] = 0;
+  const char* c = color.get() + 1;
+  int r, g, b;
+#define SCAN_BYTES(var)           \
+  colorBytes[0] = *c;             \
+  colorBytes[1] = *(c + 1);       \
+  sscanf(colorBytes, "%x", &var); \
+  c += 2;
+
+  SCAN_BYTES(r);
+  SCAN_BYTES(g);
+  SCAN_BYTES(b);
+  aColor = nsPrintfCString("rgba(%d, %d, %d, 0.2)", r, g, b);
+  return NS_OK;
+}
diff --git a/mailnews/extensions/newsblog/am-newsblog.js b/mailnews/extensions/newsblog/am-newsblog.js
--- a/mailnews/extensions/newsblog/am-newsblog.js
+++ b/mailnews/extensions/newsblog/am-newsblog.js
@@ -76,8 +76,22 @@ function serverPrettyNameOnBlur(event) {
   parent.setAccountLabel(gAccount.key, event.target.value);
   setAccountTitle();
 }
 
+function serverColorOnChange(event, reset = false) {
+  const row = parent.document.getElementById(gAccount.key);
+  if (row) {
+    if (reset) {
+      gAccount.incomingServer.color = "";
+      row.style.backgroundColor = "";
+      return;
+    }
+    const color = document.getElementById("serverColor").value;
+    gAccount.incomingServer.color = color;
+    row.style.backgroundColor = gAccount.incomingServer.RGBAcolor;
+  }
+}
+
 /**
  * Update an account's main settings title with the account name if applicable.
  *
  * @returns {void}
@@ -141,17 +155,20 @@ function setPrefs(aNode) {
 function setServerColor() {
   const colorInput = document.getElementById("serverColor");
   colorInput.value = AMUtils.serverColor;
 
-  colorInput.addEventListener("input", event =>
-    AMUtils.previewServerColor(event.target.value)
-  );
-  colorInput.addEventListener("change", event =>
-    AMUtils.updateServerColor(event.target.value)
-  );
-  document
-    .getElementById("resetColor")
-    .addEventListener("click", () => resetServerColor());
+  colorInput.addEventListener("input", event => {
+    AMUtils.previewServerColor(event.target.value);
+    serverColorOnChange(event);
+  });
+  colorInput.addEventListener("change", event => {
+    AMUtils.updateServerColor(event.target.value);
+    serverColorOnChange(event);
+  });
+  document.getElementById("resetColor").addEventListener("click", event => {
+    resetServerColor();
+    serverColorOnChange(event, true);
+  });
 }
 
 function resetServerColor() {
   document.getElementById("serverColor").value = AMUtils.defaultServerColor;
diff --git a/mailnews/extensions/newsblog/am-newsblog.xhtml b/mailnews/extensions/newsblog/am-newsblog.xhtml
--- a/mailnews/extensions/newsblog/am-newsblog.xhtml
+++ b/mailnews/extensions/newsblog/am-newsblog.xhtml
@@ -98,9 +98,15 @@
             for="serverColor"
             data-l10n-id="account-settings-color-label"
           />
           <div class="input-container">
-            <input id="serverColor" type="color" class="input-inline" />
+            <input
+              id="serverColor"
+              type="color"
+              wsm_persist="true"
+              class="input-inline"
+              prefstring="mail.server.%serverkey%.color"
+            />
             <button
               id="resetColor"
               type="button"
               data-l10n-id="account-settings-reset-color-button"
diff --git a/mailnews/mailnews.js b/mailnews/mailnews.js
--- a/mailnews/mailnews.js
+++ b/mailnews/mailnews.js
@@ -559,8 +559,9 @@ pref("mail.server.default.max_size", 50)
 pref("mail.server.default.delete_mail_left_on_server", false);
 pref("mail.server.default.valid", true);
 pref("mail.server.default.abbreviate", true);
 pref("mail.server.default.isSecure", false);
+pref("mail.server.default.color", "");
 // cleartext password. @see nsIMsgIncomingServer.authMethod.
 pref("mail.server.default.authMethod", 3);
 // @see nsIMsgIncomingServer.socketType
 pref("mail.server.default.socketType", 0);
